### PR TITLE
Update `list_tables` to account for database->schema hierarchy where applicable

### PR DESCRIPTION
**Pull request by gforsyth**
Ok, need to add tests for this, but the changes are a bit scattershot and I thought we can use this as a sounding-board for whether we want to go down this path.

The general idea here is:

1. Backends that only have a concept of `database` remain unchanged (`clickhouse`, `druid`, `impala`, `mysql`, `pyspark`, `sqlite`)
2. Backends that allow you to _interact_ with tables in other database -> schema hierarchies now have a `schema` kwarg in `list_tables` to allow fine-grained specification (formalized in `bigquery`, added for `trino`, `mssql`, and `snowflake`, and already present in `duckdb`)
3. Backends that have this hierarchy but don't allow for cross-database operations have a kwarg change from `database` to `schema` -- this unifies the arguments of `list_tables` and `table` for these backends.

For `trino`, this is quite pleasing:

```python
[ins] In [1]: import ibis

[ins] In [2]: con = ibis.trino.connect(user="user", password="", host="localhost", port=8080, data
         ...: base="memory", schema="default")

[ins] In [3]: con.list_databases()
Out[3]: ['hive', 'jmx', 'memory', 'postgresql', 'system', 'tpcds', 'tpch']

[ins] In [4]: con.list_schemas(database="tpch")
Out[4]: 
['information_schema',
 'sf1',
 'sf100',
 'sf1000',
 'sf10000',
 'sf100000',
 'sf300',
 'sf3000',
 'sf30000',
 'tiny']

[ins] In [5]: con.list_tables(database="tpch", schema="sf1")
Out[5]: 
['customer',
 'lineitem',
 'nation',
 'orders',
 'part',
 'partsupp',
 'region',
 'supplier']
```

It also adds consistency for users who are operating on Oracle or Postgres and trying to connect to tables in another schema:

```python
[ins] In [1]: import ibis
co
[ins] In [2]: con = ibis.oracle.connect(user="ibis", password="ibis", host="localhost", port=1521,
         ...:  database="ibis_testing")

[ins] In [3]: con.list_tables()
Out[3]: 
['astronauts',
 'awards_players',
 'batting',
 'diamonds',
 'functional_alltypes',
 'win']

[ins] In [4]: con.list_tables(schema="SYS", like="EXU8OPT")
Out[4]: ['EXU8OPT']

[ins] In [5]: con.table("EXU8OPT", schema="SYS")
Out[5]: 
DatabaseTable: EXU8OPT
  PARAMETER string
  VALUE     decimal(38, 0)
```

Downsides are that there's a bunch of fragmentation in the individual `list_tables` methods.
Probably can encapsulate some of this in one of the intermediate classes like `AlchemyCanCreateSchema`

xref #7329

### CODE REVIEW

**Comment by cpcloud** [comment]
5. motherduck does allow a `CREATE DATABASE` statement, making it literally and figuratively the odd duck 🦆

**Comment by cpcloud** [comment]
I think sacrificing some consistency in the `list_*` APIs is worth the decrease in conceptual distance from ibis -> backend (fewer new concepts to map from ibis to the backend)

**Comment by gforsyth** [comment]
Ok -- I'll rebase this against the changes in #7393 and then add some tests.  Then I can see about removing that deprecated `self.database` call so we can start grabbing tables more smoothly.

**Comment by cpcloud** [comment]
I went ahead and removed that `self.database()` call in #7399. It wasn't strictly necessary to fix the related bug, but I didn't want to do another partial fix.

**Comment by gforsyth** [comment]
> I went ahead and removed that `self.database()` call in #7399. It wasn't strictly necessary to fix the related bug, but I didn't want to do another partial fix.

Sounds good -- I'm taking a look at it now

File: ibis/backends/snowflake/__init__.py [hunk 1]
```diff
@@ -472,6 +474,48 @@ def list_schemas(
...
+        """
+        query = "SHOW TABLES"
+
+        if database is not None and schema is None:
+            raise com.UnsupportedArgumentError(
+                f"{self.name} cannot list tables only using `database` specifier. "
+                "Include a `schema` argument."
+            )
+        database = (
+            sg.exp.Table(catalog=database, db=schema).sql(dialect=self.name) or None
```
Comment (cpcloud) [inline ibis/backends/snowflake/__init__.py:509]: Hm, I think we may need to use an API that handles the quoting here, like `sg.table(schema, db=database, quoted=True)`

**Comment by gforsyth** [comment]
Rebased -- the changes here are not breaking, so can go into 7.1

I'm about to open a separate PR for the corresponding changes to `trino`, `snowflake`, and `bigquery` as those are breaking in the handling of keyword arguments to `list_tables`

File: ibis/backends/snowflake/__init__.py (hunk 1 above)
Comment (gforsyth) [reply ibis/backends/snowflake/__init__.py:512]: hmm, yeah.  I'll need to also add a little extra handling for passing in a dotted path to `database` to support this being backwards compatible for the moment.  At least all of the extra code will go away in 8.x

Comment (cpcloud) [reply ibis/backends/snowflake/__init__.py:512]: Hm, I'm not sure we ever supported dotted paths with `database`. Might be okay to not handle that case.

Comment (gforsyth) [reply ibis/backends/snowflake/__init__.py:512]: ugh -- well, if we use `sg.table` then passing in a dotted path to `schema` will break

File: ibis/backends/mssql/__init__.py [hunk 2]
```diff
@@ -84,6 +85,41 @@ def list_databases(self, like: str | None = None) -> list[str]:
...
         return self._scalar_query(sa.select(sa.func.schema_name()))
 
+    def list_tables(
+        self,
+        like: str | None = None,
+        database: str | None = None,
+        schema: str | None = None,
+    ) -> list[str]:
+        tablequery = sg.select("name").from_(
+            sg.exp.Table(this="tables", catalog=database, db="sys")
```
Comment (cpcloud) [inline ibis/backends/mssql/__init__.py:95]: We should use `sg.table` here. `sg.exp.Table` doesn't automatically handle quoting:

```
In [4]: import sqlglot as sg

In [5]: sg.exp.Table(this="a b c", catalog="d e f").sql("duckdb")
Out[5]: 'd e f.a b c'

In [6]: sg.table("a b c", catalog="d e f").sql("duckdb")
Out[6]: '"d e f"."a b c"'
```

Generally wherever you don't need to handle parsing, prefer `sg.table` over `sg.exp.Table`

File: ibis/backends/trino/__init__.py [hunk 3]
```diff
@@ -73,10 +74,49 @@ def current_schema(self) -> str:
...
+                ),
+                as_of="7.1",
+                removed_in="8.0",
+            )
+            database = sg.parse_one(database, into=sg.exp.Table).sql(dialect=self.name)
+        elif database is None and schema is not None:
+            database = sg.parse_one(schema, into=sg.exp.Table).sql(dialect=self.name)
+        else:
+            database = (
+                sg.table(schema, db=database, quoted=True).sql(dialect=self.name)
```
Comment (cpcloud) [inline ibis/backends/trino/__init__.py:117]: You shouldn't need to pass `quoted=True` here, Trino has sane quoting behavior.

File: ibis/backends/bigquery/__init__.py [hunk 4]
```diff
@@ -528,8 +529,48 @@ def list_databases(self, like=None):
...
+                ),
+                as_of="7.1",
+                removed_in="8.0",
+            )
+            database = sg.parse_one(database, into=sg.exp.Table).sql(dialect=self.name)
+        elif database is None and schema is not None:
+            database = sg.parse_one(schema, into=sg.exp.Table).sql(dialect=self.name)
+        else:
+            database = (
+                sg.table(schema, db=database, quoted=True).sql(dialect=self.name)
```
Comment (cpcloud) [inline ibis/backends/bigquery/__init__.py:570]: `quoted=True` shouldn't be necessary here. `sg.table` will automatically determine whether the identifier needs to be quoted.

**Review by cpcloud — CHANGES_REQUESTED** [review]
Couple of comments, but this is looking great.

**Comment by gforsyth** [comment]
Thanks for the review @cpcloud -- fixed up the bits you flagged.

**Review by cpcloud — APPROVED** [review]
Awesome!

**Pull request merged**